File: java/src/main/java/com/github/apstndb/spanvalue/SpanValue.java
package com.github.apstndb.spanvalue;

/**
 * Format Cloud Spanner types and column values.
 *
 * <p>Accepts protojson {@code Map} inputs and protobuf {@code Type}/{@code Value} objects.
 */
public final class SpanValue {
  private SpanValue() {}

  public static final String VERSION = "0.1.0";

  // Type formatting
  public static String formatType(Object typ) {
    return TypeFormat.formatType(typ, null);
  }

  public static String formatType(Object typ, TypeFormat.FormatOption option) {
    return TypeFormat.formatType(typ, option);
  }

  public static String formatTypeSimplest(Object typ) {
    return TypeFormat.formatTypeSimplest(typ);
  }

  public static String formatTypeSimple(Object typ) {
    return TypeFormat.formatTypeSimple(typ);
  }

  public static String formatTypeNormal(Object typ) {
    return TypeFormat.formatTypeNormal(typ);
  }

  public static String formatTypeVerbose(Object typ) {
    return TypeFormat.formatTypeVerbose(typ);
  }

  public static String formatTypeMoreVerbose(Object typ) {
    return TypeFormat.formatTypeMoreVerbose(typ);
  }

  public static String formatTypeVerboseAnnotationOmit(Object typ) {
    return TypeFormat.formatTypeVerboseAnnotationOmit(typ);
  }

  public static String formatTypeVerboseAnnotationPrimary(Object typ) {
    return TypeFormat.formatTypeVerboseAnnotationPrimary(typ);
  }

  // Value formatting
  public static String formatValue(Object typ, Object value, ValueFormat.FormatConfig config) {
    return ValueFormat.formatValue(typ, value, config);
  }

  public static java.util.List<String> formatRow(
      java.util.List<Object> types,
      java.util.List<Object> values,
      ValueFormat.FormatConfig config) {
    return ValueFormat.formatRow(types, values, config);
  }

  public static com.google.protobuf.Value encodeValue(Object typ, Object nativeValue) {
    return Gcvctor.encodeValue(typ, nativeValue);
  }

  public static com.google.spanner.v1.Type adaptClientType(Object clientType) {
    return ClientTypeAdapter.adapt(clientType);
  }

  public static java.util.List<String> formatResultRow(
      java.util.List<Object> types,
      java.util.List<Object> nativeValues,
      ValueFormat.FormatConfig config) {
    if (types.size() != nativeValues.size()) {
      throw new IllegalArgumentException(
          "len(types)=" + types.size() + " != len(values)=" + nativeValues.size());
    }
    java.util.List<Object> encoded = new java.util.ArrayList<>(nativeValues.size());
    for (int i = 0; i < types.size(); i++) {
      encoded.add(Gcvctor.encodeValue(types.get(i), nativeValues.get(i)));
    }
    return ValueFormat.formatRow(types, encoded, config);
  }

  public static ValueFormat.FormatConfig simpleFormatConfig() {
    return ValueFormat.simpleFormatConfig();
  }

  public static ValueFormat.FormatConfig literalFormatConfig() {
    return ValueFormat.literalFormatConfig();
  }

  public static ValueFormat.FormatConfig literalFormatConfig(Quote.LiteralQuoteConfig quote) {
    return ValueFormat.literalFormatConfig(quote);
  }

  public static ValueFormat.FormatConfig spannerCliFormatConfig() {
    return ValueFormat.spannerCliFormatConfig();
  }
}

File: ruby/lib/spanvalue.rb
# frozen_string_literal: true

require_relative 'spanvalue/version'
require_relative 'spanvalue/errors'
require_relative 'spanvalue/codes'
require_relative 'spanvalue/proto'
require_relative 'spanvalue/quote'
require_relative 'spanvalue/bytes_fmt'
require_relative 'spanvalue/float_fmt'
require_relative 'spanvalue/type_format'
require_relative 'spanvalue/format_config'

module Spanvalue
  class << self
    def format_type(typ, option = FORMAT_OPTION_SIMPLE)
      TypeFormat.format_type(typ, option)
    end

    def format_type_simplest(typ)
      TypeFormat.format_type_simplest(typ)
    end

    def format_type_simple(typ)
      TypeFormat.format_type_simple(typ)
    end

    def format_type_normal(typ)
      TypeFormat.format_type_normal(typ)
    end

    def format_type_verbose(typ)
      TypeFormat.format_type_verbose(typ)
    end

    def format_type_more_verbose(typ)
      TypeFormat.format_type_more_verbose(typ)
    end

    def format_type_verbose_annotation_omit(typ)
      TypeFormat.format_type_verbose_annotation_omit(typ)
    end

    def format_type_verbose_annotation_primary(typ)
      TypeFormat.format_type_verbose_annotation_primary(typ)
    end

    def format_type_code(code, mode = UnknownMode::VERBOSE)
      TypeFormat.format_type_code(code, mode)
    end

    def format_proto_enum(typ, mode)
      TypeFormat.format_proto_enum(typ, mode)
    end

    def format_struct_fields(fields, option)
      TypeFormat.format_struct_fields(fields, option)
    end

    def simple_format_config(null_string = '<null>')
      FormatConfigFactory.simple_format_config(null_string)
    end

    def literal_format_config(quote = nil, null_string: 'NULL')
      FormatConfigFactory.literal_format_config(quote, null_string: null_string)
    end

    def spanner_cli_format_config(null_string = 'NULL')
      FormatConfigFactory.spanner_cli_format_config(null_string)
    end

    def format_value(typ, value, config, toplevel: true)
      ValueFormat.format_value(typ, value, config, toplevel: toplevel)
    end

    def format_row(types, values, config)
      ValueFormat.format_row(types, values, config)
    end
  end
end
